dialer.go

```diff
@@ -85,3 +91,19 @@ func (w *Writer) basicDialer() (serverConn, string, error) {
 	}
 	return sc, hostname, err
 }
+
+// customDialer uses the custom dialer when the Writer was created
+// giving developers total control over how connections are made and returned.
+// Note it does not check if cdialer is nil, as it should only be referenced from getDialer.
+func (w *Writer) customDialer() (serverConn, string, error) {
+	c, err := w.cdialer(w.network, w.raddr)
```

Comment:
If we _did_ require `"custom"` as the `network`, we'd want to add a `nil` check on this before calling it. It might be safest to do that anyway.

Reply:
I figured if someone was going to take the dial functionality into their own hands, then a panic would be on them anyways (regardless of the nil check, their function could panic). That, and I was also trying to avoid the nil check such that I wouldn't have to return some new error. I had also added a caveat to the DialWithCustomDialer function, that if they passed nil it would simply act like Dial (and still not panic, since that code wouldn't be called). 

If we added a nil check, what kind of error would you want returned? A sentinel one from errors.New? A new type that implements error? This is what I was avoiding :smile:

Reply:
If there's going to be a panic, I'd rather it be in the caller's code. :wink: `srslog` code should ideally never panic even if you make a mistake in using it.

But I'd also prefer not to create a new type. I'd lean towards just a sentinel error from `errors.New`, to keep it simple.

Reply:
I added the nil check to the DialWithCustomDial function since the check needs to only be done once (not on every connection attempt) and since some of the other Dial functions make checks on their parameters before continuing.